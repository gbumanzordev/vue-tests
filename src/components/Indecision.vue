<template>
  <img v-if="image" :src="image" alt="BG" />
  <div class="bg-dark"></div>

  <div class="indecision-container">
    <input v-model="question" type="text" placeholder="Ask a question" />
    <p>Don't forget to finish your question with a question (?) mark.</p>
    <div>
      <h2>{{ question }}</h2>
      <h1 v-if="isValidQuestion">{{ answer }}</h1>
    </div>
  </div>
</template>

<script>
export default {
  name: 'Indecision',
  data() {
    return {
      question: '',
      answer: null,
      image: null,
      isValidQuestion: false,
    };
  },
  watch: {
    question(value) {
      this.isValidQuestion = false;
      console.log({ value });
      if (!value.includes('?')) {
        return;
      }
      this.isValidQuestion = true;
      this.getAnswer();
    },
  },
  methods: {
    async getAnswer() {
      try {
        this.answer = 'Thinking...';

        const { answer, image } = await fetch(
          'https://yesno.wtf/api'
        ).then((response) => response.json());

        this.answer =
          answer === 'yes' ? 'Yeah!' : answer === 'no' ? 'No!' : 'Maybe üòè';
        this.image = image;
      } catch (error) {
        this.answer = 'Unable to load from API';
        this.image = null;
        console.log(error);
      }
    },
  },
};
</script>

<style scoped>
img,
.bg-dark {
  height: 100vh;
  left: 0;
  max-height: 100%;
  max-width: 100%;
  position: fixed;
  top: 0;
  width: 100vw;
}

.bg-dark {
  background-color: rgba(0, 0, 0, 0.4);
}

.indecision-container {
  position: relative;
  z-index: 99;
}

input {
  width: 250px;
  padding: 10px 15px;
  border-radius: 5px;
  border: none;
}

input:focus {
  outline: none;
}

p {
  color: white;
  font-size: 20px;
  margin-top: 8px;
}

h1,
h2 {
  color: white;
}

h2 {
  margin-top: 150px;
}
</style>
