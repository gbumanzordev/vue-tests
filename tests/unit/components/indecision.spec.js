import { shallowMount } from '@vue/test-utils';
import Indecision from '@/components/Indecision';
import { fetchMock } from '../../utils/fetch.mock';

describe('Indecision Component', function () {
  let wrapper;
  let consoleLogSpy;
  global.fetch = jest.fn(() =>
    fetchMock({
      answer: 'yes',
      forced: false,
      image: 'https://yesno.wtf/assets/yes/2.gif',
    })
  );

  beforeEach(() => {
    wrapper = shallowMount(Indecision);
    consoleLogSpy = jest.spyOn(console, 'log');
  });

  test('should match snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('should not trigger anything when typing in input', async () => {
    const value = 'HelloWorld';
    const getAnswerSpy = jest.spyOn(wrapper.vm, 'getAnswer');

    const input = wrapper.find('input');
    await input.setValue(value);

    expect(consoleLogSpy).toHaveBeenCalledWith({ value });
    expect(getAnswerSpy).not.toHaveBeenCalled();
  });

  test('should typing question mark (?) should trigger getAnswer', async () => {
    const getAnswerSpy = jest.spyOn(wrapper.vm, 'getAnswer');
    const input = wrapper.find('input');
    await input.setValue('Imma be rich?');

    expect(getAnswerSpy).toHaveBeenCalledTimes(1);
  });

  test('should call API when getAnswer is called', async () => {
    await wrapper.vm.getAnswer();

    const img = wrapper.find('img');

    expect(img.exists()).toBeTruthy();
    expect(wrapper.vm.image).toEqual('https://yesno.wtf/assets/yes/2.gif');
    expect(wrapper.vm.answer).toEqual('Yeah!');
  });

  test('should test getAnswer when call Fails', async () => {
    fetch.mockImplementationOnce(() => Promise.reject('API is Down'));

    await wrapper.vm.getAnswer();

    const img = wrapper.find('img');

    expect(img.exists()).toBeFalsy();
    expect(wrapper.vm.answer).toBe('Unable to load from API');
  });

  afterEach(() => {
    jest.clearAllMocks();
  });
});
