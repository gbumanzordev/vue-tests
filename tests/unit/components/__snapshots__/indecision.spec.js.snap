// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indecision Component should match snapshot 1`] = `
<!--v-if-->
<div class="bg-dark"></div>
<div class="indecision-container"><input type="text" placeholder="Ask a question">
  <p>Don't forget to finish your question with a question (?) mark.</p>
  <div>
    <h2></h2>
    <!--v-if-->
  </div>
</div>
`;
